source/extensions/filters/http/cors/cors_filter.cc

```diff
@@ -194,6 +194,11 @@ Http::FilterHeadersStatus CorsFilter::encodeHeaders(Http::ResponseHeaderMap& hea
 
   headers.setInline(access_control_allow_origin_handle.handle(), latched_origin_);
 
+  // Add Vary: Origin header when origin is not "*"
+  if (latched_origin_ != "*") {
+    headers.appendCopy(Http::CustomHeaders::get().Vary, "origin");
+  }
```

Comment:
What will happen if the upstream reponse header has set `vary:origin` ?

I think you can take the `CompressorFilter::insertVaryHeader` of `/source/source/extensions/filters/http/compressor/compressor_filter.cc` as an example about how to update the vary origin.